File: FrontendApplication/frontend-project/src/app/shared/services/assignment.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { baseUrl } from "../constants/url.constant";
import { UserAssignmentDto, AssignmentDto, AssignmentAnswerDto, AssignmentResponseDto, AssignDto, AnswerDto } from "../models/assignment.model";

@Injectable({
  providedIn: 'root'
})
export class AssignmentService {
  private url: string = baseUrl + '/assignments';

  constructor(private readonly httpClient: HttpClient) {
  }

  get(): Observable<UserAssignmentDto[]> {
    return this.httpClient.get<UserAssignmentDto[]>(`${this.url}`);
  }

  getById(userAssignmentId: number): Observable<AssignmentDto> {
    return this.httpClient.get<AssignmentDto>(`${this.url}/${userAssignmentId}`);
  }

  assign(assignDto: AssignDto): Observable<any> {
    return this.httpClient.post(`${this.url}/assign`, assignDto);
  }

  startContinue(userAssignmentId: number): Observable<any> {
    return this.httpClient.put(`${this.url}/${userAssignmentId}/startContinue`, {});
  }

  finish(userAssignmentId: number): Observable<any> {
    return this.httpClient.put(`${this.url}/${userAssignmentId}/finish`, {});
  }

  answer(userAssignmentId: number, assignmentAnswerDto: AssignmentAnswerDto): Observable<AssignmentResponseDto> {
    return this.httpClient.post<AssignmentResponseDto>(
      `${this.url}/${userAssignmentId}/giveAnswer`,
      assignmentAnswerDto
    );
  }

  getAnswers(userAssignmentId: number): Observable<AnswerDto[]> {
    return this.httpClient.get<AnswerDto[]>(`${this.url}/${userAssignmentId}/answers`);
  }
}

File: APIApplication/src/main/java/org/apiapplication/services/interfaces/AssignmentService.java
package org.apiapplication.services.interfaces;

import org.apiapplication.dto.answer.AnswerDto;
import org.apiapplication.dto.assignment.*;

import java.util.List;

public interface AssignmentService {
    AssignmentDto getById(int userAssignmentId);

    List<UserAssignmentDto> get();

    void assignFunction(AssignFunctionDto assignFunctionDto);

    void assignFunctionToGroup(AssignGroupDto assignGroupDto);

    void assignMaze();

    void assignMazeToGroup(AssignGroupDto assignGroupDto);

    void startContinue(int userAssignmentId);

    void finish(int userAssignmentId);

    AssignmentResponseDto answer(int userAssignmentId, AssignmentAnswerDto assignmentAnswerDto);

    List<AnswerDto> getAnswersForAssignment(int userAssignmentId);
}
